application/src/main/java/org/togetherjava/tjbot/commands/github/GitHubReference.java

```diff
@@ -0,0 +1,168 @@
+import net.dv8tion.jda.api.entities.MessageEmbed;
+import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
+import org.apache.commons.collections4.ListUtils;
+import org.kohsuke.github.*;
+import org.togetherjava.tjbot.annotations.MethodsReturnNonnullByDefault;
+import org.togetherjava.tjbot.commands.MessageReceiverAdapter;
+import org.togetherjava.tjbot.config.Config;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.awt.*;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.List;
+
+/**
+ * GitHub Referencing feature. If someone sends #id of an issue (e.g. #207) the bot replies with an
+ * embed that contains info on the issue/pr
+ */
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+public class GitHubReference extends MessageReceiverAdapter {
+    /**
+     * The pattern used to determine whether a message is referencing an issue
+     */
+    protected static final Pattern ISSUE_REFERENCE_PATTERN = Pattern.compile("#(\\d+)");
+
+    private final Config config;
+
+    /**
+     * The repositories that are searched when looking for an issue
+     */
+    private List<GHRepository> repositories;
+
+    public GitHubReference(Config config) {
+        super(Pattern.compile(config.getGitHubReferenceChannelPattern()));
+
+        this.config = config;
+
+        try {
+            acquireRepositories();
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+    }
+
+    /**
+     * Scans all repositories mentioned in the config and adds their object representations using
+     * kohuske's API to a list
+     */
+    private void acquireRepositories() throws IOException {
+        repositories = new ArrayList<>();
+
+        GitHub githubApi = GitHub.connectUsingOAuth(config.getGitHubApiKey());
+
+        for (long repo : config.getGitHubRepositories()) {
+            repositories.add(githubApi.getRepositoryById(repo));
+        }
+    }
+
+    @Override
+    public void onMessageReceived(MessageReceivedEvent event) {
+        if (event.getAuthor().isBot()) {
+            return;
+        }
+
+        Message message = event.getMessage();
+        String content = message.getContentRaw();
+        Matcher matcher = ISSUE_REFERENCE_PATTERN.matcher(content);
+        List<MessageEmbed> embeds = new ArrayList<>();
+
+        while (matcher.find()) {
+            int id = Integer.parseInt(matcher.group(1));
+
+            try {
+                Optional<GHIssue> issue = findIssue(id);
+
+                if (issue.isPresent()) {
+                    embeds.add(generateReply(issue.get()));
+                }
+            } catch (IOException ex) {
```

Comment:
can use the optional API here. `findIssue(id).ifPresent(...)`

Reply:
then I'd need another try catch inside the lambda, which would make the code look really really ugly

Reply:
done